[Deprecation Reporting] Stub implementation

brentfulgham opened the pull request:
#### a01206bce956b19baa818f03a3f316dcdcba881a
<pre>
[Deprecation Reporting] Stub implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=245245">https://bugs.webkit.org/show_bug.cgi?id=245245</a>
&lt;rdar://problem/99988574&gt;

Reviewed by Chris Dumez.

Import the Web Platform Tests for Deprecation Reporting, create a new Experimental Feature to represent
the work, and provide a stub implementation of the report

* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/w3c-import.log: Added.
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/reporting/DeprecationReportBody.cpp: Added.
(WebCore::DeprecationReportBody::deprecationReportType):
(WebCore::DeprecationReportBody::DeprecationReportBody):
(WebCore::DeprecationReportBody::create):
(WebCore::DeprecationReportBody::type const):
(WebCore::DeprecationReportBody::id const):
(WebCore::DeprecationReportBody::anticipatedRemoval const):
(WebCore::DeprecationReportBody::message const):
(WebCore::DeprecationReportBody::sourceFile const):
(WebCore::DeprecationReportBody::lineNumber const):
(WebCore::DeprecationReportBody::columnNumber const):
(WebCore::DeprecationReportBody::createReportFormDataForViolation const):
* Source/WebCore/Modules/reporting/DeprecationReportBody.h: Copied from Source/WebCore/Modules/reporting/TestReportBody.h.
(WebCore::DeprecationReportBody::encode const):
(WebCore::DeprecationReportBody::decode):
(isType):
* Source/WebCore/Modules/reporting/DeprecationReportBody.idl: Copied from Source/WebCore/Modules/reporting/ViolationReportType.h.
* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::isVisibleToReportingObservers):
* Source/WebCore/Modules/reporting/TestReportBody.h:
* Source/WebCore/Modules/reporting/ViolationReportType.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::ReportBody&gt;&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/254634@main">https://commits.webkit.org/254634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f13c35e3d9c83f36cd7ea43c9dbf347c0bd6a9e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98934 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155572 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32687 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28164 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82004 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93343 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95289 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25970 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76500 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25920 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80849 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80730 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68913 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81308 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30443 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14811 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76177 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15743 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26420 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3252 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33641 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38688 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/78743 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32350 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34834 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17265 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+    return reportType;
+}
+
+DeprecationReportBody::DeprecationReportBody(String&& id, WallTime&& anticipatedRemoval, String&& message, String&& sourceFile)
```
WallTime is a double so no need for && or WTFMove().

brentfulgham replied:
Will fix.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+{
+}
+
+Ref<DeprecationReportBody> DeprecationReportBody::create(String&& id, WallTime&& anticipatedRemoval, String&& message, String&& sourceFile)
```
`WallTime&&` -> `WallTime`

brentfulgham replied:
Fixed.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+const String& DeprecationReportBody::id() const
+{
+    return m_id;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+WallTime DeprecationReportBody::anticipatedRemoval() const
+{
+    return m_anticipatedRemoval;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+const String& DeprecationReportBody::message() const
+{
+    return m_message;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+const String& DeprecationReportBody::sourceFile() const
+{
+    return m_sourceFile;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+std::optional<unsigned> DeprecationReportBody::lineNumber() const
+{
+    return m_lineNumber;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+
+std::optional<unsigned> DeprecationReportBody::columnNumber() const
+{
+    return m_columnNumber;
```
Should be inlined in the header.

brentfulgham replied:
Done.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+    reportBody->setString("message"_s, m_message);
+    if (!m_sourceFile.isNull()) {
+        reportBody->setString("sourceFile"_s, m_sourceFile);
+        reportBody->setInteger("lineNumber"_s, m_lineNumber ? *m_lineNumber : 0);
```
m_lineNumber.value_or(0)

brentfulgham replied:
Fixed.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,115 @@
+    if (!m_sourceFile.isNull()) {
+        reportBody->setString("sourceFile"_s, m_sourceFile);
+        reportBody->setInteger("lineNumber"_s, m_lineNumber ? *m_lineNumber : 0);
+        reportBody->setInteger("columnNumber"_s, m_columnNumber ? *m_columnNumber : 0);
```
value_or(0)

brentfulgham replied:
Fixed.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.h:
```diff
@@ -0,0 +1,106 @@
+
+class FormData;
+
+class WEBCORE_EXPORT DeprecationReportBody final : public ReportBody {
```
We should probably not mark the whole class as WEBCORE_EXPORT. We should just mark at WEBCORE_EXPORT the non-inline member functions that are actually called from outside WebCore.

brentfulgham replied:
Ah, good point. I should probably correct that in some other Report classes, but will do so in a separate patch.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.h:
```diff
@@ -0,0 +1,106 @@
+
+    const AtomString& type() const final;
+
+    const String m_id;
```
Unclear why some data members are marked as const but not others. Presumably, they all are constant?

brentfulgham replied:
Yes, that was an oversight. Fixed.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.h:
```diff
@@ -0,0 +1,106 @@
+template<typename Encoder>
+void DeprecationReportBody::encode(Encoder& encoder) const
+{
+    encoder << m_id << m_anticipatedRemoval << m_message << m_sourceFile;
```
How come we don't encode line / col numbers?

brentfulgham replied:
Oversight. I didn't have an instance where this was used, so I had left them off. I'll fix that.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.h:
```diff
@@ -0,0 +1,106 @@
+    decoder >> sourceFile;
+    if (!sourceFile)
+        return std::nullopt;
+
```
How come we don't decode line / col numbers?

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.h:
```diff
@@ -0,0 +1,106 @@
+    if (!sourceFile)
+        return std::nullopt;
+
+    return adoptRef(new DeprecationReportBody(WTFMove(*reportId), WTFMove(*anticipatedRemoval), WTFMove(*message), WTFMove(*sourceFile)));
```
No need for WTFMove() for anticipatedRemoval.

brentfulgham replied:
Fixed.

cdumez commented on Source/WebCore/Modules/reporting/DeprecationReportBody.cpp:
```diff
@@ -0,0 +1,87 @@
+
+const AtomString& DeprecationReportBody::deprecationReportType()
+{
+    static NeverDestroyed<AtomString> reportType { "deprecation"_s };
```
Since this is exposed to both main thread and workers, we likely want something like:
`static NeverDestroyed<const AtomString> reportType(MAKE_STATIC_STRING_IMPL("deprecation"));`

brentfulgham committed:
[Deprecation Reporting] Stub implementation
https://bugs.webkit.org/show_bug.cgi?id=245245
<rdar://problem/99988574>

Reviewed by Chris Dumez.

Import the Web Platform Tests for Deprecation Reporting, create a new Experimental Feature to represent
the work, and provide a stub implementation of the report

* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/META.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.js: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/idlharness.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/deprecation-reporting/w3c-import.log: Added.
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/reporting/DeprecationReportBody.cpp: Added.
(WebCore::DeprecationReportBody::deprecationReportType):
(WebCore::DeprecationReportBody::DeprecationReportBody):
(WebCore::DeprecationReportBody::create):
(WebCore::DeprecationReportBody::type const):
(WebCore::DeprecationReportBody::id const):
(WebCore::DeprecationReportBody::anticipatedRemoval const):
(WebCore::DeprecationReportBody::message const):
(WebCore::DeprecationReportBody::sourceFile const):
(WebCore::DeprecationReportBody::lineNumber const):
(WebCore::DeprecationReportBody::columnNumber const):
(WebCore::DeprecationReportBody::createReportFormDataForViolation const):
* Source/WebCore/Modules/reporting/DeprecationReportBody.h: Copied from Source/WebCore/Modules/reporting/TestReportBody.h.
(WebCore::DeprecationReportBody::encode const):
(WebCore::DeprecationReportBody::decode):
(isType):
* Source/WebCore/Modules/reporting/DeprecationReportBody.idl: Copied from Source/WebCore/Modules/reporting/ViolationReportType.h.
* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::isVisibleToReportingObservers):
* Source/WebCore/Modules/reporting/TestReportBody.h:
* Source/WebCore/Modules/reporting/ViolationReportType.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<RefPtr<WebCore::ReportBody>>::encode):
(IPC::ArgumentCoder<RefPtr<WebCore::ReportBody>>::decode):

Canonical link: https://commits.webkit.org/254634@main
